raft topology: make CDC_GENERATIONS_V3 single-partition, timeuuid-sorted

patjed41 opened the pull request:
We make the `CDC_GENERATIONS_V3` table single-partition and change the clustering key from `range_end` to `(id, range_end)`. We also change the type of `id` to `timeuuid` and ensure that a new generation always has the highest `id`. These changes allow efficient clearing of obsolete CDC generation data, which we need to prevent Raft-topology snapshots from endlessly growing as we introduce new generations over time.

All this code is protected by an experimental feature flag. It includes the definition of `CDC_GENERATIONS_V3`. The table is not created unless the feature flag is enabled.

Fixes #15163

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3539/

patjed41 commented:
I rebased this PR on #15281 because I realized there was a small conflict. It will also make working on #15323 easier since this issue depends on both of these PRs. I changed the description to indicate which commits belong to this PR.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3548/

kbr-scylla commented on cdc/generation.hh:
```diff
@@ -150,9 +150,8 @@ db_clock::time_point new_generation_timestamp(bool add_delay, std::chrono::milli
-// Works for only specific schemas: CDC_GENERATIONS_V2 (in system_distributed_keyspace)
-// and CDC_GENERATIONS_V3 (in system_keyspace).
-future<utils::chunked_vector<mutation>> get_cdc_generation_mutations(
+// Works only for the CDC_GENERATIONS_V2 schema (in system_distributed_keyspace).
```
Let's take this opportunity to fix the typo (`system_distributed_keyspace` -> `system_distributed keyspace`)

kbr-scylla commented on cdc/generation.hh:
```diff
@@ -155,4 +155,13 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations_v2(
+// using `mutation_size_threshold` to decide on the mutation sizes. All mutations have the same partition
+// key - "cdc_generation". The timestamp of each cell in each mutation is given by `mutation_timestamp`.
+//
+// Works only for the CDC_GENERATIONS_V3 schema (in system_keyspace).
```
`system_keyspace` -> `system keyspace`

kbr-scylla commented on cdc/generation.hh:
```diff
@@ -155,4 +155,13 @@ future<utils::chunked_vector<mutation>> get_cdc_generation_mutations_v2(
 
+// Translates the CDC generation data given by a `cdc::topology_description` into a vector of mutations,
+// using `mutation_size_threshold` to decide on the mutation sizes. All mutations have the same partition
+// key - "cdc_generation". The timestamp of each cell in each mutation is given by `mutation_timestamp`.
```
In the spirit of the comment for `get_cdc_generation_mutations_v2`, I would also explain how `gen_uuid` is used in the mutations (i.e. it's the first clustering key column). No need to say about `"cdc_generation"` partition key since it's unrelated to the parameters of the function and mostly an implementation detail (we have no much better way to represent single-partition tables than to invent an arbitrary partition key).

patjed41 replied:
Done.

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -2636,22 +2637,14 @@ future<service::topology> system_keyspace::load_topology_state() {
-                           CDC_GENERATIONS_V3),
+                    format("SELECT count(range_end) FROM {}.{} WHERE key = '{}' AND id = ?",
+                           NAME, CDC_GENERATIONS_V3, CDC_GENERATIONS_V3_KEY),
                     gen_uuid);
```
since we can no longer do the count sanity check, let's at least `debug` print the gen uuid and the count.

patjed41 replied:
I added that.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -393,13 +393,11 @@ topology_description limit_number_of_streams_if_needed(topology_description&& de
 }
 
-std::pair<utils::UUID, cdc::topology_description> make_new_generation_data(
+cdc::topology_description make_new_generation_description(
```
I think there's no point in keeping this function now, I'd inline it (but do as you like here.)

patjed41 replied:
`topology_description_generator` is visible only in `generation.cc`, so I can't simply inline it in `storage_service::prepare_new_cdc_generation_data`. I could make it visible, but I think we can also remove this class completely. It's used only in `make_new_generation_description` in a trivial way. We could move the body of `topology_description_generator::generate` to `make_new_generation_description` and make `topology_description_generator::get_tokens` and `topology_description_generator::create_description` static functions that take no more than two arguments.

kbr-scylla replied:
Ok, if you wish, do the refactor. (It's optional though.)

kbr-scylla commented on docs/dev/topology-over-raft.md:
```diff
@@ -187,10 +187,10 @@ CREATE TABLE system.topology (
     current_cdc_generation_timestamp timestamp static,
-    current_cdc_generation_uuid uuid static,
+    current_cdc_generation_uuid timeuuid static,
     unpublished_cdc_generations set<tuple<timestamp, id>> static,
```
btw. I noticed this now, the `id` here should be `uuid` instead, right?
(or actually, it will now become `timeuuid`?)

kbr-scylla replied:
And I see you actually didn't change the `topology()` schema in system_keyspace.cc
(it looks like you need to change all three places: `current_cdc_generation_uuid`, `unpublished_cdc_generations`, `new_cdc_generation_data_uuid`)

btw. `cdc_generation_id_v2_type` will probably deserve some better name after this (since it no longer directly corresponds to the C++ `cdc::generation_id_v2` type which is used both for old-style generation IDs where the uuids are not timeuuids, and the new-style IDs where we're storing timeuuids there)

patjed41 replied:
I fixed all these issues. I changed the name of `cdc_generation_id_v2_type` in a new last commit.

kbr-scylla commented:
I queued https://github.com/scylladb/scylladb/pull/15281, you'll have to rebase on `master` once it's merged

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -282,21 +282,21 @@ bool is_cdc_generation_optimal(const cdc::topology_description& gen, const locat
+static future<utils::chunked_vector<mutation>> get_common_cdc_generation_mutations(
         schema_ptr s,
-        utils::UUID id,
+        partition_key pkey,
```
Let's pass `pkey` through const reference, `partition_key`s are complex objects.

kbr-scylla commented:
I'd like at least one other review before merging.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3579/

gleb-cloudius commented:
> This change is backward incompatible,

You mean it isn't?

gleb-cloudius reviewed (approved):
Looks fine, but we need to be careful so that this single partition to not grow to be too large. That should be the part clean path up I guess.

patjed41 commented:
> > This change is backward incompatible,
> 
> You mean it isn't?

I clarified this statement.

patjed41 commented:
> Looks fine, but we need to be careful so that this single partition to not grow to be too large. That should be the part clean path up I guess.

I'm currently working on implementing a garbage collection mechanism -- #15323.

patjed41 commented:
v3:
- made `get_common_cdc_generation_mutations` take the `pkey` argument by const reference
- moved the declaration of `CDC_GENERATIONS_V3_KEY ` to `cdc/generation.hh` and used this constant in `cdc::get_cdc_generation_mutations_v3`
- replaced `cdc::topology_description_generator` with static functions

gleb-cloudius commented:
> > > This change is backward incompatible,
> > 
> > 
> > You mean it isn't?
> 
> I clarified this statement.

I should not comment before drinking coffee. I misread  `incompatible` as `compatible`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3618/
